### ISSUE TITLE
Files inside torrent with same file name and different size

### ISSUE THREAD

**Issue by latot** [arvidn/libtorrent#3695]
Hi, i was testing for https://github.com/arvidn/libtorrent/issues/3181, and i found something, we can create a new torrent file, the torrent have two files, with the same name, the difference the second file is the first 100mb of the first file, and the torrent is constructed correctly, so if we want share the torrent it should works, but when we load the file and check the torrent there is other result, the torrent is not completed, and in the torrent content the name of the first file is right, but the second was renamed by libtorrent, this is like "file.mp4", "file.1.mp4", so obvs how is looking for other file, not the file requested by the torrent can't be checked or shared.

i test this in v1.1.1

Thx.

**Comment by arvidn** [comment]
the size of the file is not relevant. Files aren't allowed to have the same name. Is there a question or bug report in there?

**Comment by latot** [comment]
Hi, yes, i was testing this for that other issue (https://github.com/arvidn/libtorrent/issues/3181), to can read files from other files is necessary can support this type of things, other interpretation of this issue is, read first 200mb and the again the first 100mb, i don't know if you read this comment https://github.com/arvidn/libtorrent/issues/3181#issuecomment-467746964 so i don't know if you like the idea.

Is possible implement this?, or it need to send an error?, or maybe support both, a flag to know if must be used or not.

This is to know how implement the other issue, and if this was intentional and if there is a possibility to support his.

**Comment by arvidn** [comment]
as far as I can tell, that ticket is about two separate torrents downloading/seeding from the same file. There is no direct check to prevent this, but it's error prone to do so. As long as only a single torrent *downloads* it should be fine though.

As for this ticket, I don't really know what it's about, but it sounds like you're talking about a single torrent containing multiple files with the same name (and path) which is a whole different story, and *is* in fact checked against and files are renamed.

**Comment by latot** [comment]
Hi, is right send an error when someone try rename a file with other with the same name, but it feels weird when libtorrent (alone) rename a file, and then don't found the renamed file, when exist with the original filename.

**Comment by arvidn** [comment]
what would you suggest libtorrent do with torrents with two different files with the same name (and path)?

if neither is renamed, only in very special cases would there *not* be any data corruption.

**Comment by Seeker2** [comment]
There's at least a couple very rare in-the-wild torrents out there "somewhere" that have files in the same folder that have almost-identical filenames...that differ only by CaSe.
Like "Adam.mp3" and "adam.mp3" in the same folder.
From what I've read, Linux (and Mac OS?) works fine with that...but Windows OSes has problems with that.
Shouldn't be too hard to create an example torrent under Linux to check this...

**Comment by arvidn** [comment]
right. On Mac OS, at format-time, you can configure both HFS+ and APFS to be either case insensitive or case sensitive. By default it is case insensitive though, so in general I would lump it together with Windows in this regard.

The main non-obvious decision here is: should Linux be a special case where files whose names only differing in case be allowed?

Technically, it's not really a property of the operating system, but the filesystem. On MacOS this is obvious as there is support for both kinds of volumes. I would suspect that if you mount an NTFS (and maybe samba?) drive on Linux, you also don't get case insensitivity. Similarly, I would expect that if you mount an ext4 drive on windows, you *would* get case sensitivity (at least in theory, presumably the vast majority of software on windows wouldn't handle that well).

So, really, the question is: should the filesystem we're saving to be queried to affect the renaming rules?

As far as I can tell, there are overwhelming arguments against that, but I'm open to be convinced otherwise. Here's what I have:

1. The renaming rules currently happen when a torrent is loaded into a `torrent_info`, at that point we don't know where we intend to download it to (or seed it from), if at all.

2. If a torrent is downloaded into a filesystem that's case sensitive, and them moved (via `move_storage()`) to a volume that isn't. Should the renaming rules be re-evaluated and files renamed? That's complex and possibly unintuitive.

3. If the payload of a torrent is *manually* moved from one filesystem to another, and the user expect to pick it up by adjusting the save path, it may fail if the renaming rules differ between the two drives. For example, this could be copying files from linux to windows and resuming seeding on windows.

Basically, since the renaming rules alter *compatibility* of seeding files, they better be *consistent* across filesystems and platforms and only change if there is a really good reason to change them.

libtorrent does not just rename files that have duplicate names, but also invalid UTF-8 encoding, characters that are invalid on popular filesystems and just plain invalid filenames (such as filename called "../../foobar", since "/" is a directory separator, it can't be allowed in names of files or directories).

I captured some of the reasoning in this blog post a while ago: https://blog.libtorrent.org/2014/12/filenames/

**Comment by latot** [comment]
Hi, mm, i think this have some conflicts.

The first part of the issue, if the torrent have two files with the same filename (this is how pieces are writen), with "file.txt" 200mb next "file.txt" 100mb i think libtorrent should not rename the second file to "file.1.txt", just read the torrent as it says.

The second part, i'm agree torrent should be cross-platforms, no case-sensitive, because we should not be able to seed between some filesystems, but, what should libtorrent do if there exist a torrent with that?

Would be great put this in the torrent protocol, open this in other clients too, the best for now i think is rename the filenames when the system didn't support case-sensitive, and in the other keep the original filenames, and always keep the same torrent file, change the filename but don't change the .torrent file.

For other issue would be great check the file length, ext4 for example support larger paths than NTFS or FAT, or what to do with some special characters, if all of this is already done, only left the first part of the issue.

**Comment by arvidn** [comment]
> The first part of the issue, if the torrent have two files with the same filename (this is how pieces are writen), with "file.txt" 200mb next "file.txt" 100mb i think libtorrent should not rename the second file to "file.1.txt", just read the torrent as it says.

And what would you suggest libtorrent would write to that file? Specifically the first 100 MB, that are ambiguous. Or are you suggesting it would write those parts of the file twice, and get a random mixture of pieces from two two files?

It seems pretty clear to me that doing so would not only be unintuitive (by most use cases) and also a security risk. An attacker may be able to trigger a specific interleaving of pieces to corrupt the file in specific ways.

Would you care to motivate why you think this should be the behavior? (along with what specific behaviour you're advocating).

> and always keep the same torrent file, change the filename but don't change the .torrent file.

the .torrent file is immutable, it cannot be changed without changing the info-hash (making it a different torrent).

The blog post outlines essentially what libtorrent does today. with regards to renaming.

**Comment by latot** [comment]
Hi, i was testing that, and i found that behavior, my motivation there was to see if libtorrent will read a part file or not, instead i found it rename the files, in one part is unintiuve have a fragment in the file, but personally it feels more unintuitive rename a file what "theoretically" should not be renamed, even if is unintuitive if the torrent file is right constructed all pieces and hash should be right.

Maybe instead send an error? like "Unsupported torrent format"?

**Comment by arvidn** [comment]
I don't understand that. I don't see how part files are related to this. what do you mean by a file that "should not be renamed"? why not?

> Maybe instead send an error? like "Unsupported torrent format"?

I think most users or end-user bittorrent clients would be disappointed with that behavior. The bug report would say "but uTorrent supports it".

**Comment by latot** [comment]
When we create a torrent in linux, like above, one file named file.txt and other File.txt, from the system are differents, so rename one of them when we are in windows is right to can handle the torrent.

The case when both files have exactly the same name and differents sizes would be a piece of the torrent constructed with two parts of the same file, even if is unintuitive is what the torrent file says, so in this i think libtorrent should not rename the file.

**Comment by arvidn** [comment]
ok, let's talk about the case in the second paragraph. Why is the size of the files relevant? i.e. if the two files had the same size, you *would* expect one of them to be renamed?

What bytes would you expect to be written in the overlapping region of the file? bytes from the first file or bytes from the second file?

**Comment by latot** [comment]
The actual behavior of libtorrent is rename the file (in this second case), i would expect something like this:

there is a torrent of 100mb, have two pieces of 50mb, and the structure is one file named "file" of 75mb and then again the same file obvs same name "file" but with a size of 25mb, so the first piece of the torrent, would be the first 50mb of the file, while the second piece, is the last 25mb and next the first 25mb.

Even if this a weird example i would expect this of that torrent, not rename the second file.

Obvs all hashes must match, i'm not thinking with a corrupted torrent file in mind.
And yes, there is shared data.

Thx.

**Comment by arvidn** [comment]
> so the first piece of the torrent, would be the first 50mb of the file, while the second piece, is the last 25mb and next the first 25mb.

so, the two files called "file" would effectively be concatenated on disk? so the order of the data in the file depends on the order those file entries appear in the torrent file. Would you also expect support for *creating* torrents like this?

> Obvs all hashes must match, i'm not thinking with a corrupted torrent file in mind.
> And yes, there is shared data.

As I understood your first paragraph, there would not be *any* shared data, i.e. no overlap. So, now I'm still confused.

Also, why would you expect different behaviour if the files had the same size?

**Comment by latot** [comment]
the "file" is not concatenated on disk, in some way, the filename and length says how much read/write on files, both pieces works in the same file, if you get the file from torrent you should get the file of 75mb, while indeed the torrent have 100mb, but only of how is written.
Get the first piece and half you will have all the "file" completed, so you should not need download the other data, or you can download the second piece and part of the first.

The shared data from the point of view of the torrent is the first 25mb, because is used to contruct the first mb in the first piece, and the end of the second piece.

**Comment by arvidn** [comment]
I see. so the shared data is required to be identical?

**Comment by latot** [comment]
yes, if not would be a corrupted torrent file, hashes check will never end, one piece corrupting the other...

**Comment by arvidn** [comment]
this is getting to my point. What's the point of supporting this, and risking that you end up with a corrupt file? There's no way to ensure the data is identical without downloading it (except for the case where the piece boundaries happen to be aligned perfectly).

Not to mention the added complexity of the disk I/O back-end to ensure two threads writing to the same file aren't clobbering each other.

**Comment by latot** [comment]
Yes, that is why could be better unsupport this, i was testing because in some way can be part of https://github.com/arvidn/libtorrent/issues/3181, but at least for now send an error.
Just to say, implement this, and if is used right can help to download torrents faster, if you have two torrents, and you know how construct the data of one with the other, getting differents pieces reconstruct the data can help a lot.

**Comment by Seeker2** [comment]
"The main non-obvious decision here is: should Linux be a special case where files whose names only differing in case be allowed?"

My point was there's already rare instance of torrents which do this in the wild.
Should libtorrent not support such existing torrents?

I can rename files/folders in torrents easily enough on uTorrent. I've done that when the root folder of the torrent uses non-standard characters in the name or the path+filenames would otherwise be too long. That means I could probably do the same for duplicate (except for case differences) filenames as well.

Even if the default behavior with duplicate filenames may not work, as long as there's an easy-enough workaround it won't be a problem...for me.

**Comment by arvidn** [comment]
> Edited:, to answer completely, the idea is one part, download faster reconstructing with multiples torrents, second, save space, if we need both torrents, can seed only with one and a diff file can save a lot of space.

problem is, storage is dirt cheap and plentiful, and engineering time is quite expensive and in great shortage.

If your storage needs are google-scale, it probably makes sense to trade engineering time for storage. Even if you think it's worth sequeezing out some savings here, I think a custom storage is the best approach.

**Comment by latot** [comment]
Okis, Thx!

I would suggest a warning when torrents with the case "file" "file", in the actual way we can get the "file" correctly but the torrent will never ends for now, because the file will never exist except if someone implement the custom storage and seed.

**Comment by arvidn** [comment]
I don’t understand. What, specifically, do you suggest?

**Comment by latot** [comment]
That type of torrent actually half works, will download the torrent, but will not use the torrent structure, at least one file will be renamed (the actual behavior) when this second file should not even exist.
Just to say to the user will get something similar, how torrent protocol don't support directly file offset the files will be fine, just will not be the expected.

Just like the example above, torrent of 50mb, with this structure: "file" 10mb, "file" 10mb, "file" 10mb, "file" 10mb, "file" 10mb, here torrent will only create 1 file, when is loaded libtorrent will use this "file.1" 10mb, "file.2" 10mb, "file.3" 10mb, "file.4" 10mb, "file.5" 10mb, now we have 5 equal files.
